### PR TITLE
Add series data to JSON output

### PR DESCRIPTION
**Pull request by nmattia**
This adds an `Enum` constraint to the `Series` constructor, ensuring
that we can represent `Series`' arguments when creating the report.

This allows the user to inspect the report programatically and match
benchmark results against the benchmark input.

Note that this is a breaking change for `series`; however the JSON output changes should be backwards compatible (we only add field).

Example output:

``` json
    {
      "bench_data": [
        2,
        1
      ],
      "cycles": null,
      "measurements": null,
      "garbage_collections": null,
      "time_in_nanos": 103.23967032967033,
      "bench_name": "n choose k:2:1",
      "alloc": null
    }
```

I'm not happy with the `HashMap` keys, but this is something we can change later when tackling #22 . Also the show instance for `Series` is a bit flaky, as it `show`s a `String`. Will submit a fix soon, but opening the PR already for feedback.

### CODE REVIEW

File: examples/Main.hs [hunk 1]
```diff
@@ -14,16 +14,23 @@ fact, fib :: Int -> Int
...
-          [ bench "fact" (use n >>= nf fact)
-          , bench "fib" (use n >>= nf fib)
+          [ bench "fact" (nf fact n)
+          , bench "fib" (nf fib n)
           ]
     , withSampling (timebounded (repeat 10) fiveSecs) $ bgroup "roundrip"
         [ bench "ping" (nfIO (system "ping -c1 8.8.8.8 > /dev/null")) ]
+    , series [1..4] $ \n ->
+        series [1..n] $
+          bench "n choose k" . nf (n `choose`)
```
Comment (mboes) [inline examples/Main.hs]: This is in theory not the same as `nf (uncurry choose) (n, k)`, because `choose` could well be doing part of the computation already, which will then be memoized (e.g. if there's a `let` in the body of `choose` in between abstracting the first argument and abstracting the second).

File: src/Hyperion/Analysis.hs [hunk 2]
```diff
@@ -24,37 +27,58 @@ import Hyperion.Internal
...
+
+qualData :: [Component] -> [Int]
+qualData = go []
+  where
+    go index [BenchC _txt] = index
+    go index (GroupC _txt : comps) = index <> go [] comps
+    go index (SeriesC x : comps) = go (index <> [x]) comps
     go _ _ = error "qualName: Impossible"
 
+dataOf :: Fold Benchmark (Text, [Int])
```
Comment (mboes) [inline src/Hyperion/Analysis.hs]: Poorly named: it returns more than the data.

File: src/Hyperion/Analysis.hs [hunk 3]
```diff
@@ -24,37 +27,58 @@ import Hyperion.Internal
...
+dataOf = benchmarkFoldOf (qualName &&& qualData)
+
 namesOf :: Fold Benchmark Text
-namesOf = go []
+namesOf = benchmarkFoldOf qualName
+
+benchmarkFoldOf :: ([Component] -> a) -> Fold Benchmark a
```
Comment (mboes) [inline src/Hyperion/Analysis.hs]: `Of` already means fold. You don't need to call it `benchmarkFoldOf`. It's a convention documented in `lens`. That said this function looks pretty ad hoc. Best to just inline it at the 2 call sites.

File: examples/Main.hs (hunk 1 above)
Comment (nmattia) [inline examples/Main.hs]: Good point.

**Comment by nmattia** [comment]
Split this into two PRs ( #27 and #28 ) and addressed comments.

File: src/Hyperion/Analysis.hs [hunk 4]
```diff
@@ -24,37 +28,62 @@ import Hyperion.Internal
...
+    go index (SeriesC x : comps) = go (index <> ":" <> Text.pack (show x)) comps
+    go _ _ = error "qualName: Impossible"
+
+qualData :: [Component] -> [Int]
+qualData = go []
+  where
+    go index [BenchC _txt] = index
+    go index (GroupC _txt : comps) = index <> go [] comps
+    go index (SeriesC x : comps) = go (index <> [fromEnum x]) comps
     go _ _ = error "qualName: Impossible"
```
Comment (mboes) [inline src/Hyperion/Analysis.hs]: This should say `qualData`, not `qualName`.

File: src/Hyperion/Analysis.hs [hunk 5]
```diff
@@ -24,37 +28,62 @@ import Hyperion.Internal
...
 
+-- | Return (a 'Fold' over) the benchmark name and the data associated with
```
Comment (mboes) [inline src/Hyperion/Analysis.hs]: I would say "parameters" instead of "data".

File: src/Hyperion/Analysis.hs [hunk 6]
```diff
@@ -24,37 +28,62 @@ import Hyperion.Internal
...
-        go (comps <> [SeriesC (Text.pack (show x))]) f (g x)
+        go (comps <> [SeriesC x]) f (g x)
     go comps f (WithSampling _ bk) = go comps f bk
 
     coerce :: (Contravariant f, Applicative f) => f a -> f b
     coerce = contramap (const ()) . fmap (const ())
 
 analyze
   :: Text -- ^ Benchmark name.
+  -> [Int] -- ^ Benchmark input data.
```
Comment (mboes) [inline src/Hyperion/Analysis.hs]: Benchmark parameters.

File: src/Hyperion/PrintReport.hs [hunk 7]
```diff
@@ -29,6 +28,6 @@ formatReport report =
       | x > 1e3 = show2decs (x/1e3) ++ "us"
       | otherwise = show2decs x ++ "ns"
 
-printReports :: HashMap Text Report -> IO ()
-printReports report = do
-  putDoc $ foldMap formatReport (elems report)
+printReports :: [Report] -> IO ()
```
Comment (mboes) [inline src/Hyperion/PrintReport.hs]: making the result of `printRecords` ordered rather than unordered looks to me like an orthogonal change. Should go in a separate commit, or even in a separate PR insofar as the rest of this PR doesn't build on it.

File: src/Hyperion/Report.hs [hunk 8]
```diff
@@ -29,7 +30,7 @@ data Report = Report
 makeLenses ''Report
 deriveJSON defaultOptions{ fieldLabelModifier = camelTo2 '_' . drop (length @[] "_report") } ''Report
 
-json :: UTCTime -> Maybe Text -> HashMap Text Report -> JSON.Value
+json :: UTCTime -> Maybe Text -> HashMap (Text, [Int]) Report -> JSON.Value
```
Comment (mboes) [inline src/Hyperion/Report.hs]: This calls for a record, if only to document what is going on here. Tuples are  seldom useful in Haskell.

Taking a step back, what you're doing here is saying that benchmark names aren't just stringly text anymore. They're structured things, that include a base name (of type `Text`) and zero or more benchmark parameters. If that is so, then define a datatype that says just that!

Then, you can amend the `Report` record to say:

```haskell
data Report Report
  { _reportBenchmarkId :: !BenchmarkId
  , ...
  }
```

Better yet, this makes the spurious `uncurry` occurrence in the call to `analyze` disappear, if you make `analyze` take a `BenchmarkId` as an input. That's a win for code readability, because `uncurry` is just adding distracting noise to an otherwise straightforward line in [Main.hs](https://github.com/tweag/hyperion/pull/27/files#diff-c613859236dbc1f34b63c869c01e7f26R172).

File: src/Hyperion/PrintReport.hs (hunk 7 above)
Comment (nmattia) [inline src/Hyperion/PrintReport.hs]: The idea is not to have an ordered list, but rather show that we do not care about the keys. I could go with a `Set Report` but this seems overkill.
